File: go/tests/runtime_test.go
// runtime_test.go
//
// Verifies the CircleAI.Runtime port: HostProfile helpers + StaticCapabilityProbe
// (runtime_capabilities.go), the deterministic BackendSelector branches +
// tier clamping (runtime_backend_selector.go), and the native-runtime registry +
// in-memory fetcher (runtime_native.go).

package circleai_test

import (
	"context"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestRuntime_HostProfileHelpersAndProbe(t *testing.T) {
	prof := circleai.HostProfile{
		Os:                       circleai.OSLinux,
		Arch:                     circleai.ArchX64,
		TotalPhysicalMemoryBytes: 32 << 30,
		Gpu:                      &circleai.GpuInfo{Vendor: circleai.GpuVendorNvidia, VramBytes: 8 << 30},
	}
	if !prof.Is64Bit() {
		t.Fatalf("x64 should be 64-bit")
	}
	if !prof.HasUsableGpu(2 << 30) {
		t.Fatalf("8 GiB GPU should be usable at 2 GiB threshold")
	}
	probe := circleai.NewStaticCapabilityProbe(prof)
	got, err := probe.Probe(context.Background())
	if err != nil || got.Arch != circleai.ArchX64 {
		t.Fatalf("static probe = %+v err=%v", got, err)
	}
	unk, _ := circleai.UnknownCapabilityProbe{}.Probe(context.Background())
	if unk.Os != circleai.OSUnknown || unk.CpuModel != "Unknown CPU" {
		t.Fatalf("unknown probe = %+v", unk)
	}
}

func TestRuntime_BackendSelectorBranches(t *testing.T) {
	sel := circleai.DefaultBackendSelector{}

	// Apple Silicon -> Metal.
	apple := circleai.HostProfile{Os: circleai.OSMacOS, Arch: circleai.ArchArm64, CpuModel: "M2",
		TotalPhysicalMemoryBytes: 16 << 30, Gpu: &circleai.GpuInfo{Vendor: circleai.GpuVendorApple}}
	if got := sel.Select(apple, circleai.CapabilityTier4Frontier); got.Backend != circleai.BackendMetal || got.ActualTier != circleai.CapabilityTier2Medium {
		t.Fatalf("apple selection = %+v (want Metal / Tier2 by 16GiB)", got)
	}

	// NVIDIA 24 GiB -> CUDA Tier4 (but clamped to requested Tier2).
	nv := circleai.HostProfile{Os: circleai.OSLinux, Arch: circleai.ArchX64,
		Gpu: &circleai.GpuInfo{Vendor: circleai.GpuVendorNvidia, Model: "RTX", VramBytes: 24 << 30}}
	if got := sel.Select(nv, circleai.CapabilityTier2Medium); got.Backend != circleai.BackendCuda || got.ActualTier != circleai.CapabilityTier2Medium {
		t.Fatalf("nvidia selection = %+v (want CUDA / clamped Tier2)", got)
	}

	// No accelerator -> CPU fallback, always selectable.
	cpu := circleai.HostProfile{Os: circleai.OSWindows, Arch: circleai.ArchX64, CpuModel: "Ryzen", TotalPhysicalMemoryBytes: 4 << 30}
	if got := sel.Select(cpu, circleai.CapabilityTier4Frontier); got.Backend != circleai.BackendCpu || got.ActualTier != circleai.CapabilityTier0Tiny {
		t.Fatalf("cpu fallback = %+v (want Cpu / Tier0 by 4GiB)", got)
	}

	// Ascend NPU -> Ascend.
	ascend := circleai.HostProfile{Os: circleai.OSLinux, Arch: circleai.ArchArm64,
		Npu: &circleai.NpuInfo{Vendor: circleai.NpuVendorHuaweiAscend, Model: "910"}}
	if got := sel.Select(ascend, circleai.CapabilityTier4Frontier); got.Backend != circleai.BackendAscend || got.ActualTier != circleai.CapabilityTier3Large {
		t.Fatalf("ascend selection = %+v", got)
	}
}

func TestRuntime_NativeRegistryAndFetcher(t *testing.T) {
	bundles := []circleai.NativeRuntimeBundle{
		{MnnVersion: "3.4.0", Os: circleai.OSLinux, Arch: circleai.ArchX64, Backend: circleai.BackendCpu, PrimaryURI: "u1", MnnCoreLibraryName: "libMNN.so"},
		{MnnVersion: "3.5.0", Os: circleai.OSLinux, Arch: circleai.ArchX64, Backend: circleai.BackendCpu, PrimaryURI: "u2", MnnCoreLibraryName: "libMNN.so"},
	}
	reg := circleai.NewNativeRuntimeRegistry(bundles)
	// Newest version wins.
	found, ok := reg.Find(circleai.OSLinux, circleai.ArchX64, circleai.BackendCpu)
	if !ok || found.MnnVersion != "3.5.0" {
		t.Fatalf("find newest = %+v ok=%v", found, ok)
	}
	if _, ok := reg.Find(circleai.OSWindows, circleai.ArchX64, circleai.BackendCpu); ok {
		t.Fatalf("unknown tuple must not be found")
	}

	store := circleai.NewMapRuntimeContentStore()
	store.Add(found, "/cache/linux-x64-cpu", "/cache/linux-x64-cpu/libMNN.so")
	fetcher := circleai.NewInMemoryNativeRuntimeFetcher(reg, store)

	if cached, _ := fetcher.IsRuntimeCached(context.Background(), circleai.OSLinux, circleai.ArchX64, circleai.BackendCpu); !cached {
		t.Fatalf("bundle should be cached in the store")
	}
	var lastProgress float64
	inst, err := fetcher.EnsureRuntime(context.Background(), circleai.OSLinux, circleai.ArchX64, circleai.BackendCpu, func(p float64) { lastProgress = p })
	if err != nil || inst.MnnCorePath != "/cache/linux-x64-cpu/libMNN.so" {
		t.Fatalf("ensure = %+v err=%v", inst, err)
	}
	if lastProgress != 1.0 {
		t.Fatalf("progress should end at 1.0, got %v", lastProgress)
	}
	// Unknown tuple -> error.
	if _, err := fetcher.EnsureRuntime(context.Background(), circleai.OSMacOS, circleai.ArchArm64, circleai.BackendMetal, nil); err == nil {
		t.Fatalf("unknown tuple must error")
	}
}

func TestRuntime_NativeRegistryJSON(t *testing.T) {
	json := []byte(`{"mnn_versions":[{"version":"3.5.0","bundles":[
		{"os":"Windows","arch":"X64","backend":"Cuda","url":"https://x/win.zip","sha256":"AB"},
		{"os":"BadOs","arch":"X64","backend":"Cpu","url":"https://x/bad.zip"}
	]}]}`)
	reg, err := circleai.LoadNativeRuntimeRegistryJSON(json)
	if err != nil {
		t.Fatalf("load json: %v", err)
	}
	// The BadOs entry is skipped; only the Windows/Cuda bundle survives.
	if len(reg.All()) != 1 {
		t.Fatalf("loaded %d bundles, want 1 (bad entry skipped)", len(reg.All()))
	}
	b, ok := reg.Find(circleai.OSWindows, circleai.ArchX64, circleai.BackendCuda)
	if !ok || b.MnnCoreLibraryName != "MNN.dll" {
		t.Fatalf("windows bundle = %+v ok=%v (default lib name)", b, ok)
	}
}

File: typescript/tests/runtime.test.ts
// runtime.test.ts
// Verifies the CircleAI.Runtime port: HostProfile helpers, BackendSelector table
// routing + tier clamping, the capability probes, NativeRuntimeRegistry parsing +
// newest-version lookup, and the deterministic in-memory native-runtime fetcher.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  OperatingSystemKind,
  ArchitectureKind,
  GpuVendor,
  NpuVendor,
  BackendKind,
  CapabilityTier,
  BackendSelector,
  StaticCapabilityProbe,
  UnknownCapabilityProbe,
  CapabilityProbe,
  NativeRuntimeRegistry,
  InMemoryNativeRuntimeFetcher,
  gpuInfo,
  npuInfo,
  hostProfile,
  hasUsableGpu,
  is64Bit,
  type HostProfile,
} from "../src/runtime/index";

const GIB = 1024 * 1024 * 1024;

function profile(overrides: Partial<HostProfile> = {}): HostProfile {
  return hostProfile(
    overrides.os ?? OperatingSystemKind.Linux,
    overrides.osVersion ?? "6.1",
    overrides.arch ?? ArchitectureKind.X64,
    overrides.cpuModel ?? "Test CPU",
    overrides.logicalCoreCount ?? 8,
    overrides.physicalCoreCount ?? 4,
    overrides.totalPhysicalMemoryBytes ?? 16 * GIB,
    overrides.gpu ?? null,
    overrides.npu ?? null,
    overrides.probedAt ?? new Date("2026-01-01T00:00:00Z"),
  );
}

describe("HostProfile helpers", () => {
  it("hasUsableGpu respects the VRAM floor", () => {
    assert.equal(hasUsableGpu(profile({ gpu: gpuInfo(GpuVendor.Nvidia, "x", 1 * GIB, null) })), false);
    assert.equal(hasUsableGpu(profile({ gpu: gpuInfo(GpuVendor.Nvidia, "x", 4 * GIB, null) })), true);
    assert.equal(hasUsableGpu(profile()), false);
  });

  it("is64Bit for X64/Arm64/Loong64", () => {
    assert.equal(is64Bit(profile({ arch: ArchitectureKind.X64 })), true);
    assert.equal(is64Bit(profile({ arch: ArchitectureKind.Arm64 })), true);
    assert.equal(is64Bit(profile({ arch: ArchitectureKind.X86 })), false);
  });
});

describe("BackendSelector", () => {
  const sel = new BackendSelector();

  it("routes Apple Silicon to Metal capped by unified RAM", () => {
    const p = profile({
      os: OperatingSystemKind.MacOS,
      arch: ArchitectureKind.Arm64,
      gpu: gpuInfo(GpuVendor.Apple, "M2", 0, null),
      totalPhysicalMemoryBytes: 16 * GIB,
    });
    const r = sel.select(p, CapabilityTier.Tier4_Frontier);
    assert.equal(r.backend, BackendKind.Metal);
    assert.equal(r.actualTier, CapabilityTier.Tier2_Medium); // 16 GiB unified → Tier2
  });

  it("routes NVIDIA to CUDA capped by VRAM", () => {
    const p = profile({ gpu: gpuInfo(GpuVendor.Nvidia, "RTX", 12 * GIB, "550") });
    const r = sel.select(p, CapabilityTier.Tier4_Frontier);
    assert.equal(r.backend, BackendKind.Cuda);
    assert.equal(r.actualTier, CapabilityTier.Tier3_Large); // 12 GiB VRAM → Tier3
  });

  it("routes Ascend NPU", () => {
    const p = profile({ npu: npuInfo(NpuVendor.HuaweiAscend, "910") });
    const r = sel.select(p, CapabilityTier.Tier4_Frontier);
    assert.equal(r.backend, BackendKind.Ascend);
    assert.equal(r.actualTier, CapabilityTier.Tier3_Large);
  });

  it("routes Qualcomm to OpenCL capped to Tier1", () => {
    const p = profile({ gpu: gpuInfo(GpuVendor.Qualcomm, "Adreno", 0, null) });
    const r = sel.select(p, CapabilityTier.Tier4_Frontier);
    assert.equal(r.backend, BackendKind.OpenCL);
    assert.equal(r.actualTier, CapabilityTier.Tier1_Small);
  });

  it("falls back to CPU when no accelerator is present", () => {
    const r = sel.select(profile({ totalPhysicalMemoryBytes: 16 * GIB }), CapabilityTier.Tier4_Frontier);
    assert.equal(r.backend, BackendKind.Cpu);
    assert.equal(r.actualTier, CapabilityTier.Tier1_Small); // 16 GiB CPU RAM → Tier1
    assert.match(r.rationale, /CPU SIMD backend/);
  });

  it("never upgrades above the requested tier", () => {
    const p = profile({ gpu: gpuInfo(GpuVendor.Nvidia, "RTX", 24 * GIB, null) });
    const r = sel.select(p, CapabilityTier.Tier1_Small);
    assert.equal(r.actualTier, CapabilityTier.Tier1_Small);
  });
});

describe("Capability probes", () => {
  it("StaticCapabilityProbe returns its fixed profile", async () => {
    const p = profile({ cpuModel: "Fixed" });
    assert.equal((await new StaticCapabilityProbe(p).probeAsync()).cpuModel, "Fixed");
  });

  it("UnknownCapabilityProbe returns an Unknown profile", async () => {
    const p = await new UnknownCapabilityProbe().probeAsync();
    assert.equal(p.os, OperatingSystemKind.Unknown);
    assert.equal(p.arch, ArchitectureKind.Unknown);
  });

  it("CapabilityProbe delegates to the injected inner probe", async () => {
    const p = profile({ cpuModel: "Inner" });
    assert.equal((await new CapabilityProbe(new StaticCapabilityProbe(p)).probeAsync()).cpuModel, "Inner");
  });
});

describe("NativeRuntimeRegistry", () => {
  const doc = {
    mnn_versions: [
      {
        version: "3.4.0",
        bundles: [{ os: "Windows", arch: "X64", backend: "Cpu", url: "https://example.com/a.tar.gz" }],
      },
      {
        version: "3.5.0",
        bundles: [
          {
            os: "windows",
            arch: "x64",
            backend: "cpu",
            url: "https://example.com/b.tar.gz",
            sha256: "abc",
            mnn_lib: "MNN.dll",
          },
          { os: "not-an-os", arch: "x64", backend: "cpu", url: "https://example.com/c.tar.gz" },
        ],
      },
    ],
  };

  it("parses valid bundles and skips malformed ones", () => {
    const reg = NativeRuntimeRegistry.fromJson(doc);
    assert.equal(reg.all.length, 2); // the "not-an-os" bundle is skipped
  });

  it("find returns the newest version for a tuple", () => {
    const reg = NativeRuntimeRegistry.fromJson(doc);
    const b = reg.find(OperatingSystemKind.Windows, ArchitectureKind.X64, BackendKind.Cpu);
    assert.equal(b?.mnnVersion, "3.5.0");
    assert.equal(b?.archiveSha256Hex, "abc");
  });

  it("returns an empty registry for a doc with no versions", () => {
    assert.equal(NativeRuntimeRegistry.fromJson({}).all.length, 0);
    assert.equal(NativeRuntimeRegistry.fromJson(null).all.length, 0);
  });
});

describe("InMemoryNativeRuntimeFetcher", () => {
  const reg = NativeRuntimeRegistry.fromJson({
    mnn_versions: [
      { version: "3.5.0", bundles: [{ os: "Linux", arch: "X64", backend: "Cpu", url: "https://x/a.tgz" }] },
    ],
  });

  it("resolves and caches an install for a known tuple", async () => {
    const f = new InMemoryNativeRuntimeFetcher(reg);
    assert.equal(await f.isRuntimeCachedAsync(OperatingSystemKind.Linux, ArchitectureKind.X64, BackendKind.Cpu), false);
    const install = await f.ensureRuntimeAsync(OperatingSystemKind.Linux, ArchitectureKind.X64, BackendKind.Cpu);
    assert.equal(install.bundle.mnnVersion, "3.5.0");
    assert.match(install.mnnCorePath, /libMNN\.so$/);
    assert.equal(await f.isRuntimeCachedAsync(OperatingSystemKind.Linux, ArchitectureKind.X64, BackendKind.Cpu), true);
  });

  it("rejects an unknown tuple", async () => {
    const f = new InMemoryNativeRuntimeFetcher(reg);
    await assert.rejects(() =>
      f.ensureRuntimeAsync(OperatingSystemKind.Windows, ArchitectureKind.X64, BackendKind.Cpu),
    );
  });

  it("lists the registry's bundles", () => {
    assert.equal(new InMemoryNativeRuntimeFetcher(reg).listAvailableBundles().length, 1);
  });
});
